lib/ProductOpener/Auth.pm

```diff
@@ -0,0 +1,743 @@
+	return;
+}
+
+=head2 signout_callback($request_ref)
+
+Handles the callback after successful sign-out, clears session cookie.
+
+=head3 Arguments
+
+=head4 A reference to a hash containing request information. $request_ref
+
+=head3 Return values
+
+The return URL after successful sign-out.
+=cut
+
+sub signout_callback ($request_ref) {
+	if (not(defined cookie($cookie_name))) {
+		return $formatted_subdomain;
+	}
+
+	my $state = single_param('state');
+	my %cookie_ref = cookie($cookie_name);
+	my $nonce = $cookie_ref{'nonce'};
+	if (not($state eq $nonce)) {
+		$log->info('unexpected nonce', {nonce => $nonce, expected_nonce => $state}) if $log->is_info();
+		display_error_and_exit('Invalid Nonce during OIDC logout', 500);
+	}
+
+	param('length', 'logout');
+	init_user($request_ref);
+
+	return $cookie_ref{'return_url'};
+}
+
+sub create_user_in_keycloak ($user_ref, $password) {
+	my $keycloak_users_endpoint = $oidc_options{keycloak_users_endpoint};
+	unless ($keycloak_users_endpoint) {
+		display_error_and_exit('keycloak_users_endpoint not configured', 500);
+	}
+
+	my $token = get_token_using_client_credentials();
+	unless ($token) {
+		display_error_and_exit('Could not get token to manage users with keycloak_users_endpoint', 500);
+	}
+
+	my $api_request_ref = {
+		email => $user_ref->{email},
+		emailVerified => $JSON::PP::true,    # TODO: Keep this for compat with current register endpoint?
+		enabled => $JSON::PP::true,
+		username => $user_ref->{userid},
+		credentials => [
+			{
+				type => 'password',
+				temporary => $JSON::PP::false,
+				value => $password
+			}
+		],
+		attributes => [
+			name => [$user_ref->{name}],
+			locale => [$user_ref->{initial_lc}],
+			country => [$user_ref->{initial_cc}],
+		]
+	};
+	my $json = encode_json($api_request_ref);
+
+	my $create_user_request = HTTP::Request->new(POST => $keycloak_users_endpoint);
+	$create_user_request->header('Content-Type' => 'application/json');
+	$create_user_request->header('Authorization' => $token->{token_type} . ' ' . $token->{access_token});
+	$create_user_request->content($json);
+	my $new_user_response = LWP::UserAgent::Plugin->new->request($create_user_request);
+	unless ($new_user_response->is_success) {
+		display_error_and_exit($new_user_response->content, 500);
+	}
+
+	my $get_user_request = HTTP::Request->new(GET => $new_user_response->header('location'));
+	$get_user_request->header('Content-Type' => 'application/json');
+	$get_user_request->header('Authorization' => $token->{token_type} . ' ' . $token->{access_token});
+	my $get_user_response = LWP::UserAgent::Plugin->new->request($get_user_request);
+	unless ($get_user_response->is_success) {
+		display_error_and_exit($get_user_response->content, 500);
+	}
+
+	my $json_response = $get_user_response->decoded_content(charset => 'UTF-8');
+	my @created_users = decode_json($json_response);
+	return $created_users[0];
+}
+
+sub create_user_in_product_opener ($id_token) {
```

Comment:
Is this a fallback to having users created in keycloak ? What is the intent here ?

Reply:
On the contrary. `create_user_in_product_opener` is called by `get_user_id_using_token` if the user that's signing in from Keycloak doesn't exist in Product Opener, yet: https://github.com/openfoodfacts/openfoodfacts-server/blob/b5fde3dc88ffcf41c698bc4eda84da11b9d626ae/lib/ProductOpener/Auth.pm#L282-L291

The idea is that we probably still need the `sto` files for users for some basic functionality (ie. user facets?), unless we want to rewrite _everything_ to implement this change. So if the user is new, create the `sto` file for the new user, based on data from Keycloak. Am I making any sense at all? 😅🤣

Reply:
Yes we want to keep sto file for now you are right.

So we just had to add documentation here to make it clear as you did :-)